snap-server/src/main/java/com/kyc/snap/cromulence/CromulenceSolver.java

```diff
@@ -86,10 +88,17 @@ public CromulenceSolver(DictionaryManager dictionaryManager) {
             while (bestFinalStates.size() > NUM_RESULTS)
                 bestFinalStates.pollLastEntry();
         }
-        return bestFinalStates.stream()
-            .limit(NUM_RESULTS)
-            .map(state -> new CromulenceSolverResult(state.words, state.score))
-            .collect(Collectors.toList());
+        Set<List<String>> seenWords = new HashSet<>();
+        List<CromulenceSolverResult> results = new ArrayList<>();
+        for (State state : bestFinalStates) {
+            if (seenWords.contains(state.words))
+                continue;
+            seenWords.add(state.words);
+            results.add(new CromulenceSolverResult(state.words, state.score));
+            if (results.size() == NUM_RESULTS)
+                break;
+        }
+        return results;
```

Comment:
Ideally should de-duplicate at the end of each iteration of the above while loop, but that loop is the performance bottleneck so just de-duplicating at the end here.